### ISSUE TITLE
Deprecate and repeal PEP 509 (Add a private version to dict) aka `ma_version_tag`

### ISSUE THREAD

**Issue by Fidget-Spinner** [faster-cpython/ideas#461]
See https://bugs.python.org/issue44477. PEP 509 is no longer used in CPython, we should repeat it to save space on all dictionaries using `ma_version_tag`.

I volunteer to write the PEP to repeal PEP 509.

Things the PEP will need to cover:

- [ ] Deprecation cycle
- [ ] Impact on external libraries (e.g. Cython).
- [ ] Whether we need alternatives for current use cases.

Some preliminary research tells me that Cython [does use it for cached global lookups](https://github.com/cython/cython/blob/169876872f3cb6198971a1db07e5b8a9d12b3dac/Cython/Utility/ObjectHandling.c#L1556).
We may need to offer a (private, unstable) alternative to looking up globals.

**Comment by gvanrossum** [comment]
Presumably Cython could use the same caching mechanism we now use in the code for `LOAD_GLOBAL_MODULE`? IIUC it looks at `dict->ma_keys->dk_version` instead of `ma_version_tag`.

CC @markshannon

**Comment by Fidget-Spinner** [comment]
Indeed. However, that would change the semantics of their code (their borrowed reference method would be invalid and they should follow what we do instead, which is to store an index into the dict's values array and access whatever's there).

**Comment by da-woods** [comment]
I'm not exactly sure at first look what Cython is getting from it (probably just a small optimization) but you can turn off Cython's use of this by defining `CYTHON_USE_DICT_VERSIONS=0`. So it's a very easy fix if you decide to remove it.

**Comment by da-woods** [comment]
And if CPython is no longer using it then we should probably turn it off for those versions anyway, since the "optimization" is probably useless (even if it isn't yet removed)

**Comment by vstinner** [comment]
My first motivation for PEP 509 was my https://faster-cpython.readthedocs.io/fat_python.html project which I abandonned since that time. The second motivation which convinced me to keep the PEP was the LOAD_GLOBAL optimization in Python 3.8 (it used the *dict* version).

If LOAD_GLOBAL is still optimized in Python 3.12 without using the *dict* version, I'm fine with getting ride of the *dict* version. The current PEP 659 design is more about dict *key* version, no?

**Comment by gvanrossum** [comment]
LOAD_GLOBAL in 3.11 and (so far 3.12) uses the dict keys version; see LOAD_GLOBAL_MODULE and LOAD_GLOBAL_BUILTIN in ceval.c. IIUC this deoptimizes less frequently -- checking the keys version only deopts when a new global was added, whereas checking the dict version would deopt whenever any global was given a new value. E.g. this:
```py
import sys
def f():
    global x
    for x in range(1000000):
        a = sys
```
ends up deoptimizing the old `LOAD_GLOBAL sys` instruction on each iteration because the global x was changed. The 3.11 version stays optimized because globals() has a fixed set of keys (sys, f and x).

**Comment by markshannon** [comment]
A bit more background. 
A dict object takes 8 machine words and we allocate objects as multiples of 2 words, so removing `ma_version_tag` doesn't save any memory.
Most objects do not have a dictionary, so do not need to update the version number.

However, freeing up the space means that we can add dictionary watchers without needing more space. Dictionary watchers will be important for higher tier optimizers.
So the immediate benefit is small, but in the longer term removing `ma_version_tag` is valuable.

**Comment by Fidget-Spinner** [comment]
@da-woods and @carljmeyer

Could I have both of your blessings for PEP 699 please? 
Specifically for Cython to support the backwards incompatible change, and for Cinder to support the utility provided by reusing the space.

Thank you!
https://discuss.python.org/t/pep-699-remove-private-dict-version-field-added-in-pep-509/19724

**Comment by carljm** [comment]
@Fidget-Spinner commented on the discourse thread.
